server/functions/framework/compiled_function.go

```diff
@@ -245,54 +253,68 @@ func (c *CompiledFunction) Children() []sql.Expression {
 func (c *CompiledFunction) WithChildren(children ...sql.Expression) (sql.Expression, error) {
-	return newCompiledFunctionInternal(c.Name, children, c.Functions, c.AllOverloads, c.IsOperator), nil
+	return newCompiledFunctionInternal(c.Name, children, c.OverloadTree, c.ParamPermutations, c.IsOperator), nil
 }
 
 // resolve returns an overload that either matches the given parameters exactly, or is a viable match after casting.
-// Returns a nil OverloadDeduction if a viable match is not found.
-func (c *CompiledFunction) resolve(parameters []pgtypes.DoltgresType, sources []Source) (*OverloadDeduction, []TypeCastFunction, error) {
+// Returns a nil FunctionOverloadTree if a viable match is not found.
+func (c *CompiledFunction) resolve(paramTypes []pgtypes.DoltgresType, sources []Source) (FunctionInterface, []TypeCastFunction, error) {
 	// First check for an exact match
-	exactMatch := c.Functions
-	for _, parameter := range parameters {
-		var ok bool
-		if exactMatch, ok = exactMatch.Parameter[parameter.BaseID()]; !ok {
-			break
-		}
-	}
-	if exactMatch != nil && exactMatch.Function != nil {
+	exactMatch, found := c.OverloadTree.ExactMatchForTypes(paramTypes)
+	if found {
 		return exactMatch, nil, nil
 	}
 	// There are no exact matches, so now we'll look through all of the functions to determine the best match. This is
 	// much more work, but there's a performance penalty for runtime overload resolution in Postgres as well.
 	if c.IsOperator {
-		return c.resolveOperator(parameters, sources)
+		return c.resolveOperator(paramTypes, sources)
 	} else {
-		return c.resolveFunction(parameters, sources)
+		return c.resolveFunction(paramTypes, sources)
 	}
 }
 
 // resolveFunction resolves a function according to the rules defined by Postgres.
 // https://www.postgresql.org/docs/15/typeconv-func.html
-func (c *CompiledFunction) resolveFunction(parameters []pgtypes.DoltgresType, sources []Source) (*OverloadDeduction, []TypeCastFunction, error) {
+func (c *CompiledFunction) resolveFunction(paramTypes []pgtypes.DoltgresType, sources []Source) (FunctionInterface, []TypeCastFunction, error) {
+
+	// special case: vararg funcs match a single type for all params, just need to find the right cast functions
+	if c.varArgsType != pgtypes.DoltgresTypeBaseID_Invalid {
+		casts := make([]TypeCastFunction, len(paramTypes))
+		for i := range paramTypes {
+			if polymorphicType, ok := c.varArgsType.GetRepresentativeType().(pgtypes.DoltgresPolymorphicType); ok && polymorphicType.IsValid(paramTypes[i]) {
+				casts[i] = identityCast
+			} else if casts[i] = GetImplicitCast(paramTypes[i].BaseID(), c.varArgsType); casts[i] == nil {
+				if sources[i] == Source_Constant && paramTypes[i].BaseID().GetTypeCategory() == pgtypes.TypeCategory_StringTypes {
+					casts[i] = stringLiteralCast
+				} else {
+					// can't find an appropriate cast, can't call the func
+					return nil, nil, nil
+				}
+			}
+		}
+
+		return c.OverloadTree.Function, casts, nil
+	}
```

Comment:
We shouldn't special-case this. As in the function comment, we are strictly adhering to how functions are resolved in the documentation:
https://www.postgresql.org/docs/15/typeconv-func.html
This does not adhere to that for a number of reasons.

It's straightforward to add support for `VARIADIC`, since the documentation states exactly how to handle it:
> ...the function is treated as if the array parameter were replaced by one or more occurrences of its element type, as needed to match the call.

So if the function is:
```
example_function(float4, VARIADIC int2[], text)
```
and the call is
```
example_function(1.5, 1, 2, 3, 'abc')
```
then we can expand our permutations so that we search over
```
[float4_BaseID, int2_BaseID, int2_BaseID, int2_BaseID, text_BaseID]
```

This way, we don't really have to change much of the following logic, except to know which permutation applies to which function. If the overload deducer/tree were flattened, when we could keep the matching indexes alongside the other two-dimensional slices. In `Eval()`, we'd have an additional step after casting that packages the variadic values into `[]any`, and it's passed like any other parameter.

There technically are more steps, such as choosing the non-variadic overload first if a variadic overload and non-variadic overload both match, but we could leave those as ` // TODO`s since we won't have any overloads like that for now.